### PR TITLE
feat(instrumentation-aws-sdk): upstream aws-sdk instrumentation from ext-js

### PR DESCRIPTION
**Pull request by blumamir**
This PR is for upstreaming the aws-sdk instrumentation from [this repo](https://github.com/aspecto-io/opentelemetry-ext-js/tree/master/packages/instrumentation-aws-sdk).

I copied the code here and made the required modification to match the contrib repo's tooling and conventions.

This plugin has been around for about a year and a half and has proven stable in production. it has >5K weekly downloads in npm and was actively maintained in the original repo. All issues and bug fixes along this period had been addressed and fixed.
The `aws-sdk` V2 and V3 codebase is kind of complex and was designed in a way that is not very instrumentation-friendly. I had to implement very complicated and non-trivial patches to be able to extract the needed attributes, as well as normalize the data across V2 and V3 - which can be quite hard to review and understand I suppose. Please share ideas for improvements if you have any or open discussions on things that require additional clarifications.

I am open to discuss any issue regarding the porting, in praticular:
- code and instrumentation conventions
- specification compliance and known gaps
- documentation
- testing strategy for this codebase
- breaking changes from the current published version (as this is a good chance to fix those with the new npm package name)

### CODE REVIEW

File: plugins/node/opentelemetry-instrumentation-aws-sdk/README.md
Comment (Flarna) [inline plugins/node/opentelemetry-instrumentation-aws-sdk/README.md:30]: Please remove passing the traceProvider here like we did it in other samples. Instrumentations should pick up the global `TracerProvider`.

File: plugins/node/opentelemetry-instrumentation-aws-sdk/README.md
Comment (Flarna) [inline plugins/node/opentelemetry-instrumentation-aws-sdk/README.md:24]:
```suggestion
const { AwsInstrumentation } = require('@opentelemetry/instrumentation-aws-sdk');
```

File: plugins/node/opentelemetry-instrumentation-aws-sdk/README.md
Comment (Flarna) [inline plugins/node/opentelemetry-instrumentation-aws-sdk/README.md:15]:
```suggestion
npm install --save @opentelemetry/instrumentation-aws-sdk
```

File: plugins/node/opentelemetry-instrumentation-aws-sdk/package.json
Comment (Flarna) [inline plugins/node/opentelemetry-instrumentation-aws-sdk/package.json:52]: Should we move this package also?

Comment (blumamir) [reply plugins/node/opentelemetry-instrumentation-aws-sdk/package.json:52]: Yes you are right, I need to move it as well.

**Review by willarmiros — APPROVED** [review]
Code LGTM, apart from breaking config changes discussed offline. Looks like yargs-parser dropped Node 6 and 8 support a while ago in v19.0.0: https://github.com/yargs/yargs-parser/releases/tag/v19.0.0

So if we can force the package to adopt a version before then (`18.x`) it should fix the build issues. In the future we can hopefully drop support for Node 8 altogether.

**Comment by willarmiros** [comment]
Looks like most of the linting errors can be fixed with `npm run lint:fix`

https://github.com/open-telemetry/opentelemetry-js-contrib/pull/678/checks?check_run_id=3825856708#step:5:647

File: plugins/node/opentelemetry-instrumentation-aws-sdk/src/aws-sdk.ts
Comment (willarmiros) [inline plugins/node/opentelemetry-instrumentation-aws-sdk/src/aws-sdk.ts:433]: Looks like another error is that this executor shouldn't be async... can you `await` the regionPromise outside of this executor to avoid making it async?

File: plugins/node/opentelemetry-instrumentation-aws-sdk/src/aws-sdk.ts
Comment (willarmiros) [inline plugins/node/opentelemetry-instrumentation-aws-sdk/src/aws-sdk.ts:516]: Also need to remove this

**Review by willarmiros — COMMENTED** [review]
I think you just need to remove the `ts-ignore` comments, which will obviously create new errors, but at least they'll be specific errors that we can address.

File: plugins/node/opentelemetry-instrumentation-aws-sdk/src/aws-sdk.ts
Comment (blumamir) [reply plugins/node/opentelemetry-instrumentation-aws-sdk/src/aws-sdk.ts:433]: The `_getV3MiddlewareStackResolvePatch` is sync, so I cannot do it there.
What I can do is to convert it from await to Promise.finally, but that would create an ugly code, just to resolve the lint error, which is a bummer

Comment (blumamir) [reply plugins/node/opentelemetry-instrumentation-aws-sdk/src/aws-sdk.ts]: ended up refactoring this code to use promise callbacks instead of async/await to calm down the linter.
The code is less idiomatic and harder to read. we might want to refactor this one day

**Review by willarmiros — COMMENTED** [review]
Great to see the CI checks passing, unfortunate we had to do some acrobatics to get lint passing but it still LGTM and should be ready for review by @open-telemetry/javascript-approvers if they want to.

The only thing blocking from merging is adding the new packages to the [release-please config](https://github.com/open-telemetry/opentelemetry-js-contrib/blob/main/release-please-config.json)

File: plugins/node/opentelemetry-instrumentation-aws-sdk/doc/sqs.md
Comment (Rauno56) [inline plugins/node/opentelemetry-instrumentation-aws-sdk/doc/sqs.md:30]:
```suggestion
1. The SDK is calling a single callback for the messages batch, and it's not straightforward to understand when each individual message processing starts and ends (and set the context correctly for cascading spans).
```

File: plugins/node/opentelemetry-instrumentation-aws-sdk/package.json
Comment (Rauno56) [reply plugins/node/opentelemetry-instrumentation-aws-sdk/package.json:4]: There are other places that need that fix as well...

File: packages/opentelemetry-propagation-utils/package.json
Comment (blumamir) [reply packages/opentelemetry-propagation-utils/package.json:23]: Changed. these utils are not only for aws, they are general for messaging systems

File: plugins/node/opentelemetry-instrumentation-aws-sdk/README.md
Comment (willarmiros) [reply plugins/node/opentelemetry-instrumentation-aws-sdk/README.md:62]: yeah I'd always be in favor of adding them later if customers ask for them rather than adding them eagerly if they're awkward to support in v3.

**Comment by mattfysh** [comment]
Hey @blumamir - thanks for this great contrib!

I had a question about using the table name as "DB_NAME" attribute, I wonder if this should be "DynamoDB" as shown in this screenshot: https://docs.aws.amazon.com/xray/latest/devguide/images/scorekeep-PUTrules-timeline.png

The table name would then be stored alongside the operation name, and rendered by the Xray UI in the segment description as "OperationName: TableName"

On a loosely related note, do you know if there is any documentation covering how Xray decides what to render as the grey description text (to the right of the coloured bar in the timeline)? Thanks!

**Comment by blumamir** [comment]
> I had a question about using the table name as "DB_NAME" attribute, I wonder if this should be "DynamoDB" as shown in this screenshot: https://docs.aws.amazon.com/xray/latest/devguide/images/scorekeep-PUTrules-timeline.png

Currently, `db.name` is filled with the table name, and `db.system` contains the "dynamodb" value. I agree that maybe the table name is more appropriate in an attribute like `db.table`, but the spec only has [`db.sql.table`](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/database.md#call-level-attributes-for-specific-technologies) but dyanmodb is not an sql database.

> The table name would then be stored alongside the operation name, and rendered by the Xray UI in the segment description as "OperationName: TableName"

Not sure what you are suggesting here. Are you suggesting to add the table name in the span name? or to concat the table name into the `db.operation` attribute?

Regarding Xray, I have no experience with it. maybe @willarmiros  or @NathanielRN  can help

**Comment by willarmiros** [comment]
> On a loosely related note, do you know if there is any documentation covering how Xray decides what to render as the grey description text (to the right of the coloured bar in the timeline)? Thanks!

I don't believe this is documented since it's more of an internal implementation detail - the closest documentation we have is [this on viewing traces](https://docs.aws.amazon.com/xray/latest/devguide/xray-console.html#xray-console-traces). However I can say that basically the gray text attempts to represent the operation performed, and depends on the `namespace` of a subsegment, and the attributes of that subsegment. For instance, if the namespace is `aws`, then the text is `aws.operation`, and for some cases it will also include a resource like for DynamoDB it will be `aws.operation: aws.table_name`. If the namespace is `remote`, the gray text will be something like `Remote: http.request.method`.

**Pull request merged**
